src/lib.rs

```diff
@@ -430,6 +432,311 @@ impl<T> OptionExt<T> for Option<T> {
+    /// use snafu::{FutureExt, Snafu};
+    ///
+    /// #[derive(Debug, Snafu)]
+    /// enum Error {
+    ///     Authenticating { user_name: String, user_id: i32, source: ApiError },
+    /// }
+    ///
+    /// fn example() -> impl Future<Item=i32, Error=Error> {
+    ///     another_function()
+    ///         .map(|v| v + 10)
+    ///         .eager_context(Authenticating { user_name: "admin", user_id: 42 })
+    /// }
+    ///
+    /// # type ApiError = std::io::Error;
+    /// fn another_function() -> impl Future<Item=i32, Error=ApiError> {
+    ///     /* ... */
+    /// # Ok(42).into_future()
+    /// }
+    /// ```
+    fn eager_context<C, E2>(self, context: C) -> EagerContextFuture<Self, C, E, E2>
+    where
+        Context<E, C>: Into<E2>;
+
+    /// Extend a `Future`'s error with lazily-generated context-sensitive
+    /// information which when polled immediately converts an error to another.
+    ///
+    /// This is most useful when using `Future`'s combinators and when
+    /// the final `Error` type is already constrained.
+    ///
+    /// ```rust
+    /// # extern crate futures;
+    /// use futures::Future;
+    /// # use futures::IntoFuture;
+    /// use snafu::{FutureExt, Snafu};
+    ///
+    /// #[derive(Debug, Snafu)]
+    /// enum Error {
+    ///     Authenticating { user_name: String, user_id: i32, source: ApiError },
+    /// }
+    ///
+    /// fn example() -> impl Future<Item=i32, Error=Error> {
+    ///     another_function()
+    ///         .map(|v| v + 10)
+    ///         .with_eager_context(|| Authenticating {
+    ///             user_name: "admin".to_string(),
+    ///             user_id: 42,
+    ///         })
+    /// }
+    ///
+    /// # type ApiError = std::io::Error;
+    /// fn another_function() -> impl Future<Item=i32, Error=ApiError> {
+    ///     /* ... */
+    /// # Ok(42).into_future()
+    /// }
+    /// ```
+    ///
+    /// Note that this *may not* be needed in many cases because the
+    /// [`From`](std::convert::From) implementation generated by the
+    /// macro will call [`Into::into`](std::convert::Into::into) on
+    /// each field.
+    fn with_eager_context<C, F, E2>(self, context: F) -> WithEagerContextFuture<Self, C, E, F, E2>
+    where
+        F: FnOnce() -> C,
+        Context<E, C>: Into<E2>;
+}
+
+/// Future for the `context` combinator, changing the error type of a future.
+///
+/// See the `FutureExt::context` method for more details.
+pub struct ContextFuture<A, C> {
+    context: Option<C>,
+    future: A,
+}
+
+impl<A: futures::Future, C> futures::Future for ContextFuture<A, C> {
+    type Item = A::Item;
+    type Error = Context<A::Error, C>;
+
+    fn poll(&mut self) -> Result<futures::Async<Self::Item>, Self::Error> {
+        self.future.poll().map_err(|error| Context {
+            error,
+            context: self
+                .context
+                .take()
+                .expect("cannot poll ContextFuture twice"),
```

Comment:
I know it's part of futures, but hate having to do this. 😿

Reply:
This should be something like "cannot poll ContextFuture after it has resolved", as it's totally fine to poll it twice before that.